dace/codegen/targets/fpga.py

```diff
@@ -160,68 +163,242 @@ def on_target_used(self) -> None:
         # Right before finalizing code, write FPGA context to state structure
         self._frame.statestruct.append('dace::fpga::Context *fpga_context;')
 
-    def generate_state(self, sdfg, state, function_stream, callsite_stream):
-        """Generate a kernel that runs all connected components within a state
-           as concurrent dataflow modules."""
-
+    def _kernels_subgraphs(self, graph: Union[dace.sdfg.SDFGState,
+                                              ScopeSubgraphView],
+                           dependencies: dict):
+        '''
+            Finds subgraphs of an SDFGState or ScopeSubgraphView that correspond to kernels.
+            This is done by looking at the kernel ID associated to each node ('_kernel' field).
+            :param graph, the state/subgraph to consider
+            :param dependencies: a dictionary containing for each kernel ID, the IDs of the kernels from which it
+                depends on
+            :return a list of tuples (subgraph, kernel ID) topologically ordered according kernel dependencies.
+        '''
+        from dace.sdfg.scope import ScopeSubgraphView
+
+        if not isinstance(graph, (dace.sdfg.SDFGState, ScopeSubgraphView)):
+            raise TypeError(
+                "Expected SDFGState or ScopeSubgraphView, got: {}".format(
+                    type(graph).__name__))
+
+        subgraphs = collections.defaultdict(
+            set)  # {kernel_id: {nodes in subgraph}}
```

Comment:
Why not use a list instead of a set? I don't think you can ever have duplicates, and then you can maintain the order for later

Reply:
Because I can find the same node multiple times (e.g. Access Node)

Reply:
I don't think you can? Aren't you iterating over `all_nodes_recursive`? This should never return the same node more than once

Reply:
You are right!